src/backend/functions/orchestrator/lib/index.js

```diff
@@ -102,35 +114,106 @@ function createDeleteParams({mediaUrl, tasksTableName}) {
-const terminated = R.curry(deleteItem =>
-   handle(o(deleteItem, createDeleteParams), ({MediaUrl: mediaUrl}) => ({mediaUrl}))
-);
+// terminated :: ({k:v} -> Promise {k: v}) -> {k: v} -> {k: v} -> Promise {k:v}
+const terminated = R.curry(deleteItem => o(deleteItem, createDeleteParams));
+
+// errorUpdate :: ({k: v} -> Promise {k: v}) -> {k: v} -> Promise {k: v}
+const errorUpdate = R.curry((updateItem, {tasksTableName, mediaUrl, taskArn}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: mediaUrl
+      },
+      UpdateExpression: 'ADD Retries :val SET TaskStatus = :status, TaskArn = :task',
+      ExpressionAttributeValues: {
+         ':val': 1,
+         ':status':  'WAITING',
+         ':task': taskArn
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params)
+       .then(({Attributes}) => ({...Attributes, tasksTableName}))
+});
+
+const unrecoverableError = R.curry((updateItem, {tasksTableName, mediaUrl}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: mediaUrl
+      },
+      UpdateExpression: 'SET TaskStatus = :status',
+      ExpressionAttributeValues: {
+         ':status':  'UNRECOVERABLE_ERROR'
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params);
+});
 
-function error(ecs, ddb) {}
+// error :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> Promise {k:v}
+const error = R.curry((ecs, updateItem) => composeP(errorUpdate(updateItem), startTranscription(ecs)));
 
+function isUnrecoverableError(retries, retryThreshold, taskStatus) {
+   return retries > retryThreshold && !R.includes(taskStatus, ['TERMINATING', 'TERMINATED']);
+}
 
-// handler :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> ({k: v} -> {k: v} -> Promise {k: v})
+// handler :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> ({k: v} -> Promise {k: v})
 module.exports = (ecs, ddb, env) => {
+   const update = params => ddb.update(params).promise();
+
    const handlers = {
-      WAITING: waiting(ecs, params => ddb.update(params).promise()),
+      WAITING: waiting(ecs, update),
       INITIALIZING: noop,
       PROCESSING: noop,
       TERMINATING: terminating(ecs),
       TERMINATED: terminated(params => ddb.delete(params).promise()),
-      ERROR: error
+      ERROR: error(ecs, update),
+      UNRECOVERABLE_ERROR: noop
    };
 
-   return (event, context) => {
+   const normalisedEnv = convertEnvVars(env);
+
+   return event => {
+
       const promises = R.map(record => {
          const image = converter.unmarshall(R.path(['dynamodb', 'NewImage'], record));
-         const handler = R.defaultTo(noop, handlers[image.TaskStatus]);
-         return handler(env, image);
+
+         if(R.isEmpty(image)) return Promise.resolve({}); // this means we've a delete event
+
+         const params = R.mergeRight(normalisedEnv, convertImage(image));
+
+         const {mediaUrl, retries = 0, taskStatus, tasksTableName, retryThreshold = '3'} = params;
```

Comment:
how come this isn't just an int (and Number in the CFN template)? just wondering if there's a risk of getting NaN here which will cause `isUnrecoverableError` to always return `false`

Reply:
`retries` come from the the DynamoDb record so is already cast to a number by `unmarshall`, `retryThreshold` comes from an environment variable so it's always a string, which is why I do `parseInt` on it below.

Reply:
i mean in the scenario where I fat finger the env var and put "BLAH" or something, then `parseInt` will return NaN?Changing the parameter type to `Number` in the CFN template will protect against this. It'll still come in as a string here so this code will remain the same.

Reply:
Ah sorry, completely misunderstood. Yeah, make sense.